src/node.cc

```diff
@@ -2118,8 +2119,8 @@ extern "C" void node_module_register(void* m) {
     mp->nm_flags = NM_F_LINKED;
     mp->nm_link = modlist_linked;
     modlist_linked = mp;
-  } else {
-    modpending = mp;
+  } else if (mods.count(mp->nm_modname) == 0) {
+    mods[mp->nm_modname] = mp;
```

Comment:
I think we need a stronger key because this also breaks those projects which use two different versions of the same module. So, like,
```
project
  |
  +-dep1
  |   |
  |   +-native-addon1(@2.2.0)
  |
  +-native-addon1(@1.2.0)
```
The distinction between the two modules with identical `nm_modname` is the path from which they are loaded. So we should save the absolute path of the module we are about to load to a global variable declared next to `modpending` right before calling `Dlib.Open()`, and then use that path as the key here.

So, like
```C++
static node_module* modpending;
std::string modpending_filename;
```
and then
```C++
mods[modpending_filename.c_str()] = mp;
```
and
```C++
modpending_filename = std::string(*filename);
Dlib dlib(*filename, flags);
```
In general, I have reservations about attempting to duplicate what the library loader is doing, because that code has very likely been refined to a very high degree over the course of the decades.

Reply:
You are absolutely right, the basename is not unambiguous enough.

I can change it as you suggest - in fact we don't even need to save the pending filename, we can just he the pending `node_module *` as we do today and use its file name as the key in the `std::unordered_map`.

Reply:
@frutiger we must also ensure that two different file names which refer to the same file in reality are treated correctly. I'm not sure to what extent the JS side of `require()` takes care of resolving files in this way.

Reply:
@frutiger it might be worth also taking a look at the JS side, and maybe adding some more resolution steps to portion of the the code path following the decision to load a native module.

Reply:
@frutiger I guess we should at least follow symlinks.

Reply:
Ideally, the JS loader logic would cause the resulting call to `dlopen()` to always return a new module, and run the constructors. However, this means that we will have reproduced `dlopen()` decision as to whether to load a library or increment the refcount of an existing one. I'm pretty sure we haven't done that.

The decision made by `dlopen()` as to whether to load a new library, thereby running its constructors, or incrementing the reference count of an existing library, thereby not running its constructors, is what we're trying to replicate here. The code for that decision is very old, very well-vetted, and may even be platform-dependent. This is why I said I was reluctant to attempt to reproduce that code.

In an ideal world we'd have `dlreopen()` or the `RTLD_REOPEN` flag which would do exactly what we need. But alas.

Reply:
I guess @bnoordhuis' modification whereby it falls back to looking for a well-known symbol is a way to kick-start the addon if it fails to register. But, like you said, this won't work for existing modules. *sigh*